gh-132975: Improve Remote PDB interrupt handling

godlygeek opened the pull request:
Fix several problems in how Ctrl-C works in the Remote PDB client:

- In some places we would capture a KeyboardInterrupt and proxy it to the remote, but in other cases the KeyboardInterrupt wound up killing the client (notably a Ctrl-C caught while it was printing a message would kill it, and you could easily get into that situation with a PDB command that spams messages, like `for x in itertools.count(): x`). Fix this by explicitly blocking SIGINT signals except for the times when we're prepared to handle them and ready to send them on to the remote.
- On Windows, it's not possible to interrupt the `sockfile.readline()` call made while waiting for the next message from the remote. Work around this by using `selectors` to listen for updates on either of two streams - either a message from the remote, _or_ a signal arriving and being written to a `socketpair` by our signal handler.
- On Unix, send a SIGINT signal directly to the remote process when we want to interrupt it. This isn't applicable on Windows, where we can't send SIGINT to another process. Sending a SIGINT signal has all the disadvantages listed in https://docs.python.org/3/library/signal.html#note-on-signal-handlers-and-exceptions but it has the advantage of matching what happens for normal non-remote PDB, and the additional advantage of being able to interrupt IO.

Making this a draft PR for now to check whether the selectors and socketpair usage need us to skip tests on additional platforms.

<!-- gh-issue-number: gh-132975 -->
* Issue: gh-132975
<!-- /gh-issue-number -->

godlygeek committed:
Only allow KeyboardInterrupt at specific times

Because our goal is to proxy interrupt signals to the remote process,
it's important that we only accept those signals at moments when we're
ready to handle them and send them on. In particular, we're prepared to
handle them when reading user input for a prompt, and when waiting for
the server to send a new prompt after our last command. We do not want
to accept them at other times, like while we're in the middle of
printing output to the screen, as otherwise a

    while True: "Hello"

executed at the PDB REPL can't reliably be Ctrl-C'd: it's more likely
that the signal will arrive while the client is printing than while it's
waiting for the next line from the server, and if so nothing will be
prepared to handle the `KeyboardInterrupt` and the client will exit.

godlygeek committed:
Have _PdbClient work with the socket directly

Previously we worked with a file object created with `socket.makefile`,
but on Windows the `readline()` method of a socket file can't be
interrupted with Ctrl-C, and neither can `recv()` on a socket. This
refactor lays the ground work for a solution this this problem.

godlygeek committed:
Allow interrupting socket reads on Windows

Since a socket read can't be directly interrupted on Windows, use the
`selectors` module to watch for either new data to be read from that
socket, or for a signal to arrive (leveraging `signal.set_wakeup_fd`).
This allows us to watch for both types of events and handle whichever
arrives first.

godlygeek committed:
Use a SIGINT to interrupt the remote on Unix

This unfortunately cannot be applied to Windows, which has no way to
trigger a SIGINT signal in a remote process without running code inside
of that process itself.

gaogaotiantian reviewed (commented):
Okay so `_readline` is to make it possible for the client to be interrupted while receiving data from the server. You have a `_handle_sigint` outside of each `_readline` call which enables `SIGINT`. What if the interrupt happens before your `try ... finally ...` block in `_readline`? `old_wakeup_fd` would not be restored I think? Not sure if other evil thing could happen.

As far as I can tell, if you want to set your signal handler in `_readline()` anyway, you should just restore the signal handling in `readline()` - you wouldn't need to wrap every call with `_handle_sigint` and less racing concerns.

I see that you put some effort to achieve a "delayed" interrupt compared "ignored" ones on Windows. Is that critical? We will have much less code if we treat them the same. I think interrupting at ignored region should be the rare case anyway right?

I'm not an expert in signal, if we have multiple `SIGINT` during the blackout region, will we trigger multiple handlers?

`signal.default_int_handler` is not documented, but it is used a few times in stdlib. If that's the only handler we are going to use for `_handle_sigint`, should we just put it into the function and call it `_unblock_sigint` so it's obvious that it has the opposite effect of `_block_sigint`?

Also, we need a news entry for this PR. This is a user observable change (even though it in alpha phase).

godlygeek committed:
Handle a ValueError for operations on a closed file

It seems like this is raised explicitly in some cases.

godlygeek committed:
Use a single complex signal handler for the PDB client

godlygeek committed:
Add a news entry

gaogaotiantian commented on Lib/pdb.py:
```diff
@@ -2677,7 +2678,7 @@ def _send(self, **kwargs):
             self._sockfile.write(json_payload.encode() + b"\n")
             self._sockfile.flush()
-        except OSError:
+        except (OSError, ValueError):
```
What's the new `ValueError` about?

godlygeek replied:
I hit it while testing something unrelated. Seems like we can either get an `OSError` if the socket has been disconnected but `.close()` hasn't been called on the sockfile yet, or a `ValueError` if `.close()` has been called.

I think that case happens if we try to send a message after `detach()` has been called, though I'm not sure how exactly I triggered it. Maybe sending a `KeyboardInterrupt` that arrived after a `q` was processed?

gaogaotiantian commented on Lib/pdb.py:
```diff
@@ -2957,8 +2962,40 @@ def _send(self, **kwargs):
+
     def read_command(self, prompt):
-        reply = input(prompt)
+        with self._sigint_raises_keyboard_interrupt():
```
Do we have to put this exactly around `input()`? We should be able to do just around `read_command` right? `KeyboardInteruupt` is handled outside of the function. Or this has something to do with the one-shot thing?

godlygeek replied:
I think it would be fine to put it around `read_command`, if you prefer that. I tried to scope it as narrowly as possible to make it as clear as possible why we're doing it, but I can't see any harm to doing it outside instead.

gaogaotiantian replied:
My concern is that, it's not impossible that in the future we add more `input()` in this function, and we need to remember to add the context around it (maybe we forgot what that's about). It's probably safer to just wrap the whole function.

gaogaotiantian replied:
Do you have strong opinions here?

godlygeek replied:
I was thinking about this while adding the `self.multiline_block` in #133387. In that case it's fine if we don't reset it after the `input()` call, since we'll just wind up setting it again before the next one. But I originally thought about doing a context manager that sets `self.multiline_block = True` on entry, and sets it to False on exit. That's harder to do if we make the change you're proposing here, because it would be possible for the `KeyboardInterrupt` to be raised inside the `finally` block of the context manager, causing the assignment to not happen.

That is, this would allow the `__exit__` of `_is_multiline_block` to be interrupted by a `KeyboardInterrupt`:
```py
with self._sigint_raises_keyboard_interrupt():
    input(prompt)
    ...
    with self._is_multiline_block(True):
        input(continuation_prompt)
```
This wouldn't:
```py
with self._sigint_raises_keyboard_interrupt():
    input(prompt)
...
with self._is_multiline_block(True):
    with self._sigint_raises_keyboard_interrupt():
        input(continuation_prompt)
```

I'm pretty confident that we want to scope where `KeyboardInterrupt` can occur absolutely as narrowly as possible. I think this suggestion is more likely to introduce bugs than prevent them.

If we require it for each `input()` call and you forget, the impact is that some `input()` call isn't interruptible - which should be easy to notice while testing. If we wrap it around the entire `read_command()` function call, the impact is that some things will be interruptible that we didn't want to be, and that'll be hard to notice when testing because it's deeply dependent upon timing.

gaogaotiantian replied:
Okay I think that makes sense. Then could we wrap `input()` in a method? Something like `self.interruptible_input` or even `self.read_input()`, so in the future we know that this is the method to read input from users intead of raw input and we wouldn't miss the context. Also now that you mentioned it, it might make sense to make multiline state a context.

godlygeek replied:
> Then could we wrap `input()` in a method?

Done

> it might make sense to make multiline state a context.

I thought about it on the other PR, and convinced myself that's more complex and has no advantage. The only thing that reads from it is the completer, the only time the completer runs is underneath input(), so just setting it explicitly before each call to input() is sufficient.

I did move setting it into the new method that's wrapping input(), though, so that it's always set to the correct value whenever we prompt for input.

gaogaotiantian commented on Lib/pdb.py:
```diff
@@ -3009,11 +3047,70 @@ def readline_completion(self, completer):
+                    # Restore the old wakeup fd if we installed a new one
+                    if old_wakeup_fd is not sentinel:
+                        signal.set_wakeup_fd(old_wakeup_fd)
+                    self.signal_read = self.signal_write = None
```
Should this line be in the outmost `finally`?

godlygeek replied:
It doesn't really matter much, and we could also just remove this line entirely - once we've restored the old wakeup fd, nothing will ever use either of those sockets again, whether we set them to `None` or not.

But I do prefer cleaning up after myself, so you're right, and I'll swap it to the outer `finally`.

gaogaotiantian commented on Lib/pdb.py:
```diff
@@ -3031,11 +3128,20 @@ def cmdloop(self):
-            " (Use 'cont' to resume)."
-        )
-        sys.remote_exec(self.pid, self.interrupt_script)
+        if hasattr(signal, "pthread_kill"):
```
Now that we don't even use `pthread_kill`, I think we could have better options to detect platform. I think the most commonly used is `sys.platform == "win32"`.

godlygeek replied:
What we want here isn't really Windows vs not-Windows, it's Unix vs not-Unix.

For now `sys.platform == "win32"` is good enough, but if we ever got `sys.remote_exec()` and remote PDB ported to WASI or something weird like that, that guard might need updating.

I'll update it to `sys.platform == "win32"`, though, and leave that problem for another day.

gaogaotiantian reviewed (commented):
I think this approach is easier to understand than the previous one. I have a few comments. Also do we still have any coverage for the new feature?

godlygeek committed:
Update the comment to explain why we catch two exception types

godlygeek committed:
Swap signal_read/signal_write resetting to the outer finally block

godlygeek committed:
Use os.kill() on every platform but Windows

godlygeek committed:
Use `ExitStack` to reduce nesting in `attach`

godlygeek committed:
Use a thread to manage interrupts on Windows

gaogaotiantian reviewed (commented):
I'm okay with the overall approach. A few minor points:
1. Could you test that the current code works with commands? I was a bit worried about how you use commands there - will it work when you expand those and write as a file?
2. We have a few random functions that are only about remote pdb lying around now. It may not be a huge issue, but I was wondering if it makes more sense to put them into `_PdbServer` and `PdbClient` as `classmethod`s - so it's clear that they belong to remote pdb and we know which end we should call it.

godlygeek committed:
Use the signal handling thread approach on all platforms

Let the caller of `_connect` tell us whether to start the thread.

godlygeek committed:
Make all _connect arguments keyword-only

This gives us greater flexibility to reorder things or add things in the
future without breaking backwards compatibility.

gaogaotiantian commented on Lib/pdb.py:
```diff
@@ -92,8 +93,11 @@
+import threading
 import _colorize
 
+from contextlib import ExitStack
```
I think at this point, we are importing 3 different things from `contextlib` and we used 3 lines - that's unnecessary. We could either put them in the same line, or just import `contextlib` and use atttributes.

gaogaotiantian reviewed (commented):
Overall I think it's good for now. Need to resovle the conflict though.

godlygeek committed:
Merge remote-tracking branch 'upstream/main' into improve_remote_pdb_interrupt_handling

godlygeek committed:
One line for contextlib imports

godlygeek committed:
Add some tests for handling SIGINT in the PDB client

godlygeek committed:
Switch back to os.kill() on Unix

godlygeek committed:
Wrap input() calls in a function

gaogaotiantian reviewed (approved):
Yeah let's land this today and polish it later if necessary.

pablogsal committed:
Merge branch 'main' into improve_remote_pdb_interrupt_handling
